File: sqlite3/cplus/create/sqlite3_create.cpp
// --------------------------------------------------------------------
/*
	sqlite3_create.cpp

					Oct/21/2014

*/
// --------------------------------------------------------------------
#include	<iostream>
#include	<fstream>
#include	<map>

#include	<string.h>
#include	<boost/lexical_cast.hpp>

#include	<sqlite3.h>

using namespace std;
using namespace boost;

typedef map<string,string> Unit;

// --------------------------------------------------------------------
extern	void dict_display_proc (map <string,Unit> dict_aa);
extern	map <string,Unit> dict_append_proc
	(map <string,Unit> dict_aa,string id,string name,
	int population,string date_mod);

// --------------------------------------------------------------------
static int callback (void *NotUsed, int argc, char *argv[],char *azColName[])
{
	int it;

	for (it=0; it<argc; it++)
		{
		cout << argv[it] << "\t";
		}

	cout << "\n";

	return 0;
}

// --------------------------------------------------------------------
static map <string,Unit >  data_prepare_proc ()
{
	map <string,Unit> dict_aa;

	dict_aa = dict_append_proc (dict_aa,"t0711","郡山",34795,"1986-10-07");

	dict_aa = dict_append_proc (dict_aa,"t0712","会津若松",15982,"1986-2-12");
	dict_aa = dict_append_proc (dict_aa,"t0713","白河",73421,"1986-3-28");
	dict_aa = dict_append_proc (dict_aa,"t0714","福島",63519,"1986-2-21");

	dict_aa = dict_append_proc (dict_aa,"t0715","喜多方",83971,"1986-1-5");
	dict_aa = dict_append_proc (dict_aa,"t0716","二本松",73149,"1986-5-22");
	dict_aa = dict_append_proc (dict_aa,"t0717","いわき",46785,"1986-7-17");
	dict_aa = dict_append_proc (dict_aa,"t0718","相馬",54618,"1986-3-04");
	dict_aa = dict_append_proc (dict_aa,"t0719","須賀川",74823,"1986-5-9");

	return	dict_aa;
}

// --------------------------------------------------------------------
void table_insert_proc (sqlite3 *db,map <string,Unit> dict_aa)
{
	char *zErrMsg = 0;
	map <string,Unit>:: iterator it = dict_aa.begin ();

	while (it != dict_aa.end ())
		{
		Unit unit_aa = (*it).second;

		string id =  (*it).first;
		string name = unit_aa["name"];
		string population = unit_aa["population"];
		string date_mod = unit_aa["date_mod"];

		char str_sql[200];
	sprintf (str_sql,
"insert into cities (id,name,population,date_mod) values ('%s','%s',%s,'%s')" ,
	id.c_str (),name.c_str (),population.c_str (),date_mod.c_str ());

//	cout << str_sql << "\n";

	int rc = sqlite3_exec (db,str_sql, callback, 0, &zErrMsg);
		it++;
		}
}

// --------------------------------------------------------------------
int main (int argc,char *argv[])
{
	cerr << "*** 開始 ***\n";

	char file_sqlite3[160];

	strcpy (file_sqlite3,argv[1]);

	cerr << "\tfile_sqlite3 = " << file_sqlite3 << "\n";

	map <string,Unit> dict_aa = data_prepare_proc ();

	cout << "要素数：" << (unsigned int)dict_aa.size() << endl;

	dict_display_proc (dict_aa);

	sqlite3 *db;
	char *zErrMsg = 0;

	int rc;

	rc = sqlite3_open (file_sqlite3, &db);

	if (rc)
		{
		fprintf (stderr,"Can't open database: %s\n",sqlite3_errmsg(db));
		sqlite3_close(db);

		exit (1);
		}

	rc = sqlite3_exec (db,"drop table cities", callback, 0, &zErrMsg);

	if( rc!=SQLITE_OK )
		{
		fprintf(stderr, "SQL error: %s\n", zErrMsg);
		sqlite3_free(zErrMsg);
		}

	string sql_str = "create table cities (id varchar(10), name varchar(20), population int, date_mod date)";

	rc = sqlite3_exec (db,sql_str.c_str (), callback, 0, &zErrMsg);

	if( rc!=SQLITE_OK )
		{
		fprintf(stderr, "SQL error: %s\n", zErrMsg);
		sqlite3_free(zErrMsg);
		}

	table_insert_proc (db,dict_aa);

	sqlite3_close(db);

	cerr << "\n*** 終了 ***\n";
}

// --------------------------------------------------------------------

File: sqlite3/ocaml/create/sqlite3_create.ml
(* ------------------------------------------------------------------- *)
(*
	sqlite3_create.ml

					Nov/21/2014

*)
(* ------------------------------------------------------------------- *)
open Printf
open Sqlite3

exception Dummy
(* ------------------------------------------------------------------- *)
let sql_execute_proc = fun db sql_str -> 
	try
	match
	exec db sql_str ~cb:(fun _ _ -> print_endline "FOUND!"; raise Dummy)
	with
	| Rc.OK -> print_endline "OK"
	| _ -> assert false
	with xcp -> print_endline (Printexc.to_string xcp);;
	
(* ------------------------------------------------------------------- *)
let sql_insert_proc = fun db id_in name_in population_in date_mod_in -> 
  let sql_str = sprintf
 "insert into cities (id, Name, Population, date_mod) values (\'"
 ^ id_in ^ "\',\'" ^ name_in ^ "\',"
 ^ string_of_int population_in ^ ",\'" ^ date_mod_in ^ "\')" in
(*    print_endline sql_str *)
    sql_execute_proc db sql_str

(* ------------------------------------------------------------------- *)
let () =
	print_endline "*** 開始 ***";
	printf "%s\n" Sys.argv.(1);

	let db = db_open Sys.argv.(1) in

	let sql_aa = sprintf "drop table if exist cities" in
		sql_execute_proc db sql_aa ;

	let sql_bb = sprintf "create TABLE cities ( id varchar(10) NOT NULL PRIMARY KEY, name text, population int, date_mod text)" in
		sql_execute_proc db sql_bb ;

	sql_insert_proc db "t0711" "郡山" 51936 "1995-9-21";
	sql_insert_proc db "t0712" "会津若松" 48517 "1995-3-12";
	sql_insert_proc db "t0713" "白河" 72395 "1995-7-24";
	sql_insert_proc db "t0714" "福島" 62561 "1995-5-22";
	sql_insert_proc db "t0715" "喜多方" 42597 "1995-9-14";
	sql_insert_proc db "t0716" "二本松" 37528 "1995-3-15";
	sql_insert_proc db "t0717" "いわき" 84519 "1995-7-19";
	sql_insert_proc db "t0718" "相馬" 21587 "1995-8-18";
	sql_insert_proc db "t0719" "須賀川" 41642 "1995-10-25";
	print_endline "*** 終了 ***"
(* ------------------------------------------------------------------- *)
